examples/job-shop-scheduling/jobShopScheduler.py

```diff
@@ -0,0 +1,268 @@
+    JobShopScheduler.csp.
+    """
+    return "{task.job}_{task.position},{time}".format(**locals())
+
+
+class Task:
+    def __init__(self, job, position, machine, duration):
+        self.job = job
+        self.position = position
+        self.machine = machine
+        self.duration = duration
+
+    def __repr__(self):
+        return ("{{job: {job}, position: {position}, machine: {machine}, duration:"
+                " {duration}}}").format(**vars(self))
+
+
+class KeyList:
+    """A wrapper to an array. Used for passing the key of a custom object to the bisect function.
+
+    Note: bisect function does not let you choose an arbitrary key, hence this class was created.
+    """
+
+    def __init__(self, array, key_function):
+        self.array = array  # An iterable
+        self.key_function = key_function  # Function for grabbing the key of a given item
+
+    def __len__(self):
+        return len(self.array)
+
+    def __getitem__(self, index):
+        item = self.array[index]
+        key = self.key_function(item)
+        return key
+
+
+class JobShopScheduler:
+    def __init__(self, job_dict, max_time=None):
+        self.tasks = []
+        self.last_task_indices = []
+        self.max_time = max_time
+        self.csp = dwavebinarycsp.ConstraintSatisfactionProblem(dwavebinarycsp.BINARY)
+
+        # Populates self.tasks and self.max_time
+        self._process_data(job_dict)
+
+    def _process_data(self, jobs):
+        """Process user input into a format that is more convenient for JobShopScheduler functions.
+        """
+        # Create and concatenate Task objects
+        tasks = []
+        last_task_indices = [-1]    # -1 for zero-indexing
+        total_time = 0  # total time of all jobs
+
+        for job_name, job_tasks in jobs.items():
+            last_task_indices.append(last_task_indices[-1] + len(job_tasks))
+
+            for i, (machine, time_span) in enumerate(job_tasks):
+                tasks.append(Task(job_name, i, machine, time_span))
+                total_time += time_span
+
+        # Update values
+        self.tasks = tasks
+        self.last_task_indices = last_task_indices[1:]
+
+        if self.max_time is None:
+            self.max_time = total_time
+
+    def _add_one_start_constraint(self):
+        """self.csp gets the constraint: A task can start once and only once
+        """
+        for task in self.tasks:
+            task_times = {get_label(task, t) for t in range(self.max_time)}
+            self.csp.add_constraint(sum_to_one, task_times)
+
+    def _add_precedence_constraint(self):
+        """self.csp gets the constraint: Task must follow a particular order.
+         Note: assumes self.tasks are sorted by jobs and then by position
+        """
+        valid_edges = {(0, 0), (1, 0), (0, 1)}
+        for i, current_task in enumerate(self.tasks[:-1]):
+            next_task = self.tasks[i + 1]
```

Comment:
Since you don't really need index in the rest of the loop, an idiomatic way would be to iterate over zipped shifted lists:
```
for current_task, next_task in zip(self.tasks, self.tasks[1:]):
```